src/main/java/org/generationcp/commons/util/FileNameGenerator.java

```diff
@@ -1,83 +1,261 @@
-		final String extension = fileExt.contains(".") ? fileExt : "." + fileExt;
-		final boolean isAppend = isAppendExtension.length == 1 ? isAppendExtension[0]: true;
-		final StringBuilder fileName = new StringBuilder();
-		fileName.append(origFinalName);
-		fileName.append("_");
-		fileName.append(FileNameGenerator.getUserNameTimeStamp());
-		if (isAppend) {
-			fileName.append(extension);
-		}
-		return FileNameGenerator.truncateIfNecessary(fileName.toString(), FileNameGenerator.getFileNameMaxSize(isAppend, extension));
+	public static String generateFileName(final Project project, final ToolName toolName, final String extension,
+										  final String... fileNames) {
+		final String directory = installationDirectoryUtil.getOutputDirectoryForProjectAndTool(project, toolName);
+		return generateFileName(directory, extension, fileNames);
 	}
 
-	public static String generateFileName(final String origFinalName) {
-
-		final int maxSize;
-		String fileExt = "";
-		String oFName = origFinalName;
-		if (origFinalName.contains(".")) {
-			final String oFileName[] = origFinalName.split("\\.");
-			oFName = oFileName[0];
-			if (oFileName.length >= 2) {
-				// Get the Last Group
-				fileExt = oFileName[oFileName.length - 1];
+	public static String generateFileName(final String directory, final String extension, final String... fileNames) {
+		final int maxLengthPerName = MAX_SIZE_NAME / fileNames.length;
+		final StringBuilder filename = new StringBuilder();
+		Optional<String> hasExt = Optional.empty();
+		for (final String name : fileNames) {
+			final Optional<String> nameExt = hasExtension(name);
+			if (nameExt.isPresent()) {
+				hasExt = nameExt;
+			}
+			filename.append(truncateIfNecessary(removeExtension(name), maxLengthPerName, true));
+		}
+		final String ext;
+		if (StringUtils.isEmpty(extension)) {
+			if (hasExt.isPresent()) {
+				ext = preppend(hasExt.get(), ".");
+			} else {
+				ext = extension;
 			}
-			maxSize = FileNameGenerator.getFileNameMaxSize(true, fileExt);
 		} else {
-			maxSize = FileNameGenerator.MAX_SIZE_NAME;
+			ext = preppend(extension, ".");
+		}
+		String genFileName = getFormattedFileName(filename.toString()) + ext;
+		genFileName = fileNameChecker(directory, genFileName);
+		return truncateIfNecessary(genFileName, MAX_SIZE, false);
+	}
+
+	public static String generateFileName(final String filename) {
+		final String genName = FileNameGenerator.getFormattedFileName(filename);
+		return genName;
+	}
+
+	public static String removeExtension(final String filename) {
+		final Optional<String> extension = hasExtension(filename);
+		if (extension.isPresent()) {
+			final String ext = preppend(extension.get(), ".");
+			final String woFilename = filename.replace(ext, "");
+			return woFilename;
+		}
+		return filename;
+	}
+
+	public static String removeSuffixIfApplicable(final String filename) {
```

Comment:
This feels like a hack to me. It seems like a complex logic just to generate a filename. Why would a filename have duplicate date/time/username in the first place?

Reply:
Hi @abatac , 
This is used when appending another file name eg. _SummaryStat to the generated file name, this ensures that the file will have the suffix username_date_time